File: backend/app/db/redis.py
import redis.asyncio as redis

from app.core.config import settings

redis_client: redis.Redis | None = None


async def connect_redis() -> None:
    global redis_client
    redis_client = redis.from_url(settings.REDIS_URL, decode_responses=True)
    await redis_client.ping()


async def close_redis() -> None:
    if redis_client:
        await redis_client.aclose()


def get_redis() -> redis.Redis:
    if redis_client is None:
        raise RuntimeError("Redis has not been initialized")
    return redis_client

File: backend/src/db/redis.ts
import { Redis } from "ioredis";
import { config } from "../core/config.js";
import { logger } from "../core/logger.js";

let redisClient: Redis | null = null;

export async function connectRedis(): Promise<Redis | null> {
  if (redisClient) {
    return redisClient;
  }

  try {
    redisClient = new Redis(config.REDIS_URL, {
      lazyConnect: true,
      maxRetriesPerRequest: 2,
      retryStrategy(times) {
        if (times > 3) {
          return null; // Stop retrying if Redis is not available
        }
        return Math.min(times * 500, 2000);
      },
    });

    redisClient.on("error", (err) => {
      logger.warn({ error: err.message }, "redis_error");
    });

    await redisClient.connect();
    await redisClient.ping();
    logger.info("redis_connected");
    return redisClient;
  } catch (error) {
    logger.warn({ error: (error as Error).message }, "redis_connection_warning");
    return null;
  }
}

export async function closeRedis(): Promise<void> {
  if (redisClient) {
    await redisClient.quit();
    redisClient = null;
    logger.info("redis_closed");
  }
}

export function getRedis(): Redis | null {
  return redisClient;
}
